src/robot/running/arguments/typeconverters.py

```diff
@@ -389,61 +389,165 @@ class ListConverter(TypeConverter):
+            if TypeConverter.converter_for(self.nested_types[0],
+                                           languages=self.languages):
+                for i, elem in enumerate(converted_list) :
+                    converter = TypeConverter.converter_for(self.nested_types[0],
+                                                            languages=self.languages)
+                    if converter:
+                        elem_info = f"List[{type_name(self.nested_types[0])}]"
+                        converted_list[i] = converter.convert(elem_info, elem,
+                                                              explicit_type)
+        return converted_list
 
-    def _convert(self, value, explicit_type=True):
-        return self._literal_eval(value, list)
+    _convert = _non_string_convert
 
 
 @TypeConverter.register
 class TupleConverter(TypeConverter):
     type = tuple
     type_name = 'tuple'
     value_types = (str, Sequence)
+    aliases = ('tuple',)
 
-    def _non_string_convert(self, value, explicit_type=True):
-        return tuple(value)
-
-    def _convert(self, value, explicit_type=True):
-        return self._literal_eval(value, tuple)
+    def __init__(self, type_, custom_converters=None, languages=None):
+        super().__init__(type_, languages=languages)
+        self.nested_types = getattr(type_, '__args__', ())
 
+    def _non_string_convert(self, value, explicit_type=True):
+        if isinstance(value, str):
+            converted_tuple = self._literal_eval(value, tuple)
+        elif is_list_like(value):
+            converted_tuple = tuple(value)
+        else:
+            raise ValueError
+        if self.nested_types:
+            tmp_list = list(converted_tuple)
+            for i, elem in enumerate(converted_tuple):
+                converter = TypeConverter.converter_for(self.nested_types[i],
+                                                        languages=self.languages)
+                if converter:
+                    types = ', '.join([type_name(t) for t in self.nested_types])
+                    tmp_list[i] = converter.convert(f"Tuple[{types}]", elem,
+                                                    explicit_type)
+            converted_tuple = tuple(tmp_list)
+        return converted_tuple
+
+    _convert = _non_string_convert
 
 @TypeConverter.register
 class DictionaryConverter(TypeConverter):
     type = dict
     abc = Mapping
+    abc = (Mapping, MutableMapping)
     type_name = 'dictionary'
-    aliases = ('dict', 'map')
+    aliases = ('dict', 'dictionary', 'map')
     value_types = (str, Mapping)
 
-    def _non_string_convert(self, value, explicit_type=True):
-        if issubclass(self.used_type, dict) and not isinstance(value, dict):
-            return dict(value)
-        return value
+    def __init__(self, type_, custom_converters=None, languages=None):
+        super().__init__(type_, languages=languages)
+        self.nested_types = getattr(type_, '__args__', ())
```

Comment:
Custom converters need to be passed forward.

Here we could validate that `len(self.nested_types)` is either 0 or 2 (like with `list` we could check it's 0 or 1). There wouldn't be a need to look that again later anymore.